File: compiler/bench/strrev.c
// expected: 774920
// Reverse 10K strings of length 100
// Checksum: sum of first char after reverse (which was last char before)

#include <stdio.h>

char buf[128];

void fill_buf(int n) {
    for (int i = 0; i < 100; i++) {
        buf[i] = 'A' + (n % 26);
    }
}

void reverse_buf(void) {
    for (int i = 0; i < 50; i++) {
        char t = buf[i];
        buf[i] = buf[99 - i];
        buf[99 - i] = t;
    }
}

int main(void) {
    long sum = 0;
    for (int i = 0; i < 10000; i++) {
        fill_buf(i);
        reverse_buf();
        sum += buf[0];
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/strrev.fs
\ expected: 774920
\ Reverse 10K strings of length 100
\ Checksum: sum of first char after reverse (which was last char before)

create buf 128 allot

: fill-buf ( n -- ) 100 0 do dup 26 mod 65 + buf i + c! loop drop ;
: reverse-buf 100 2 / 0 do buf i + c@ buf 99 i - + c@ buf i + c! buf 99 i - + c! loop ;
: bench-strrev ( -- sum )
  0
  10000 0 do
    i fill-buf
    reverse-buf
    buf c@ +
  loop ;

: main bench-strrev . cr ;
